binread/src/binread_impls.rs

```diff
@@ -117,55 +117,50 @@ impl<C: Copy + 'static, B: BinRead<Args = C>> BinRead for Vec<B> {
     }
 }
 
-macro_rules! binread_array_impl {
-    ($($size:literal),*$(,)?) => {
-        $(
-            impl<C: Copy + 'static, B: BinRead<Args = C> + Default> BinRead for [B; $size] {
-                type Args = B::Args;
+impl<C: Copy + 'static, B: BinRead<Args = C>, const N: usize> BinRead for [B; N] {
+    type Args = B::Args;
 
-                fn read_options<R: Read + Seek>(reader: &mut R, options: &ReadOptions, args: Self::Args) -> BinResult<Self> {
-                    #[cfg(feature = "debug_template")]
-                    {
-                        let pos = reader.seek(SeekFrom::Current(0))?;
-                        let type_name = core::any::type_name::<B>().rsplitn(1, "::").nth(0).unwrap();
-
-                        if let Some(name) = options.variable_name {
-                            binary_template::write_vec_named(
-                                options.endian, pos, type_name, $size, name
-                            );
-                        } else {
-                            binary_template::write_vec(options.endian, pos, type_name, $size);
-                        }
-                    }
+    fn read_options<R: Read + Seek>(
+        reader: &mut R,
+        options: &ReadOptions,
+        args: Self::Args,
+    ) -> BinResult<Self> {
+        #[cfg(feature = "debug_template")]
+        {
+            let pos = reader.seek(SeekFrom::Current(0))?;
+            let type_name = core::any::type_name::<B>().rsplitn(1, "::").nth(0).unwrap();
 
-                    #[cfg(feature = "debug_template")]
-                    let options = &ReadOptions {
-                        dont_output_to_template: true,
-                        ..*options
-                    };
-
-                    let mut arr: [B; $size] = Default::default();
-                    for elem in arr.iter_mut() {
-                        *elem = BinRead::read_options(reader, options, args)?;
-                    }
-                    Ok(arr)
-                }
+            if let Some(name) = options.variable_name {
+                binary_template::write_vec_named(options.endian, pos, type_name, N, name);
+            } else {
+                binary_template::write_vec(options.endian, pos, type_name, N);
+            }
+        }
 
-                fn after_parse<R>(&mut self, reader: &mut R, ro: &ReadOptions, args: B::Args)-> BinResult<()>
-                    where R: Read + Seek,
-                {
-                    for val in self.iter_mut() {
-                        val.after_parse(reader, ro, args)?;
-                    }
+        #[cfg(feature = "debug_template")]
+        let options = &ReadOptions {
+            dont_output_to_template: true,
+            ..*options
+        };
 
-                    Ok(())
-                }
-            }
-        )*
+        let mut arr: [B; N] = unsafe { core::mem::MaybeUninit::uninit().assume_init() };
```

Comment:
It's only safe to .assume_init() immediately on a `MaybeUninit<[T; N]>` if `T` is itself `MaybeUninit<U>` so the above line should look something like:

```rust
let mut arr: [MabyeUninit<B>; N] = unsafe { core::mem::MaybeUninit::uninit().assume_init() };
```

then each item should be assigned `MaybeUninit::new(some_val)`, and then the entire array should be transmuted to `[B; N]`

Reply:
Fixed that, restored the old implementation, and put my new code behind a feature gate "const_generics".
Previous Rust compatibility is preserved while Rust 1.51+ users can benefit from const generics :)

Feel free to merge this PR or squash all commits into one.